Firestore/Source/Core/FSTView.mm

```diff
@@ -242,57 +243,73 @@ - (FSTViewDocumentChanges *)computeChangesWithDocuments:(FSTMaybeDocumentDiction
-        if (docsEqual) {
-          [changeSet addChange:[FSTDocumentViewChange
-                                   changeWithDocument:newDoc
-                                                 type:FSTDocumentViewChangeTypeMetadata]];
-        } else {
+      if (!docsEqual) {
+        if (![self shouldWaitForSyncedDocument:newDoc oldDocument:oldDoc]) {
           [changeSet addChange:[FSTDocumentViewChange
                                    changeWithDocument:newDoc
                                                  type:FSTDocumentViewChangeTypeModified]];
-        }
+          changeApplied = YES;
 
-        if (lastDocInLimit && self.query.comparator(newDoc, lastDocInLimit) > 0) {
-          // This doc moved from inside the limit to after the limit. That means there may be some
-          // doc in the local cache that's actually less than this one.
-          needsRefill = YES;
+          if (lastDocInLimit && self.query.comparator(newDoc, lastDocInLimit) > 0) {
+            // This doc moved from inside the limit to after the limit. That means there may be
+            // some doc in the local cache that's actually less than this one.
+            needsRefill = YES;
+          }
         }
+      } else if (oldDocHadPendingMutations != newDocHasPendingMutations) {
+        [changeSet
+            addChange:[FSTDocumentViewChange changeWithDocument:newDoc
+                                                           type:FSTDocumentViewChangeTypeMetadata]];
+        changeApplied = true;
       }
+
     } else if (!oldDoc && newDoc) {
       [changeSet
           addChange:[FSTDocumentViewChange changeWithDocument:newDoc
                                                          type:FSTDocumentViewChangeTypeAdded]];
+      changeApplied = true;
     } else if (oldDoc && !newDoc) {
       [changeSet
           addChange:[FSTDocumentViewChange changeWithDocument:oldDoc
                                                          type:FSTDocumentViewChangeTypeRemoved]];
+      changeApplied = true;
+
       if (lastDocInLimit) {
         // A doc was removed from a full limit query. We'll need to re-query from the local cache
         // to see if we know about some other doc that should be in the results.
         needsRefill = YES;
       }
     }
+
+    if (changeApplied) {
+      if (newDoc) {
+        newDocumentSet = [newDocumentSet documentSetByAddingDocument:newDoc];
+        if (newDoc.hasLocalMutations) {
+          newMutatedKeys = newMutatedKeys.insert(key);
+        } else {
+          newMutatedKeys = newMutatedKeys.erase(key);
+        }
+      } else {
+        newDocumentSet = [newDocumentSet documentSetByRemovingKey:key];
+        newMutatedKeys = newMutatedKeys.erase(key);
+      }
+    }
   }];
   if (self.query.limit) {
-    // TODO(klimt): Make DocumentSet size be constant time.
-    while (newDocumentSet.count > self.query.limit) {
+    for (long i = newDocumentSet.count - self.query.limit; i > 0; --i) {
```

Comment:
This is fine. If you want to iterate forward (which makes this feel more natural to me), you could do this instead:

```
long numberOfDocumentsToRemove = newDocumentSet.count - self.query.limit;
for (long i = 0; i<numberOfDocumentsToRemove; i++) {
```